src/serializer/ResidualHeapVisitor.js

```diff
@@ -400,26 +394,80 @@ export class ResidualHeapVisitor {
     }
   }
 
-  visitValueSet(val: ObjectValue): void {
-    let kind = val.getKind();
+  visitValueWeakMap(val: ObjectValue): void {
+    invariant(val.getKind() === "WeakMap");
+    let entries = val.$WeakMapData;
 
-    let entries;
-    if (kind === "Set") {
-      entries = val.$SetData;
-    } else {
-      invariant(kind === "WeakSet");
-      entries = val.$WeakSetData;
-    }
     invariant(entries !== undefined);
     let len = entries.length;
 
+    for (let i = 0; i < len; i++) {
+      this._withUnrelatedScope(this.scope, () => {
+        let entry = entries[i];
+        let key = entry.$Key;
+        let value = entry.$Value;
+
+        if (key !== undefined && value !== undefined) {
+          let fixpoint_rerun = () => {
+            let progress;
+            if (this.values.has(key)) {
+              progress = true;
+              this.visitValue(key);
+              this.visitValue(value);
+            } else {
+              progress = false;
+              this._withUnrelatedScope(this.scope, fixpoint_rerun);
+            }
+            return progress;
+          };
+          return fixpoint_rerun();
+        }
+      });
+    }
+  }
+
+  visitValueSet(val: ObjectValue): void {
+    invariant(val.getKind() === "Set");
+
+    let entries = val.$SetData;
+    invariant(entries !== undefined);
+
+    let len = entries.length;
     for (let i = 0; i < len; i++) {
       let entry = entries[i];
       if (entry === undefined) continue;
       this.visitValue(entry);
     }
   }
 
+  visitValueWeakSet(val: ObjectValue): void {
+    invariant(val.getKind() === "WeakSet");
+
+    let entries = val.$WeakSetData;
+    invariant(entries !== undefined);
+
+    let len = entries.length;
+    for (let i = 0; i < len; i++) {
+      this._withUnrelatedScope(this.scope, () => {
```

Comment:
This outer withUnrelatedScope call should not be needed anymore (and if it was needed, it should go after `entry !== undefined` check).